WalletWasabi/WabiSabi/Client/WabiSabiHttpApiClient.cs

```diff
@@ -67,15 +71,52 @@ public async Task<RoundState[]> GetStatusAsync(CancellationToken cancellationTok
 		public Task ReadyToSignAsync(ReadyToSignRequestRequest request, CancellationToken cancellationToken) =>
 			SendAndReceiveAsync<ReadyToSignRequestRequest>(RemoteAction.ReadyToSign, request, cancellationToken);
 
+		private async Task<HttpResponseMessage> SendWithRetriesAsync(RemoteAction action, string jsonString, CancellationToken cancellationToken)
+		{
+			var errors = new List<Exception>();
+
+			var start = DateTime.Now;
+
+			for (var attempt = 0; attempt < MaxRetries; attempt++)
+			{
+				try
+				{
+					using StringContent content = new(jsonString, Encoding.UTF8, "application/json");
+
+					// Any transport layer errors will throw an exception here.
+					var response = await _client.SendAsync(HttpMethod.Post, GetUriEndPoint(action), content, cancellationToken).ConfigureAwait(false);
+
+					var totalTime = DateTime.Now - start;
+
+					if (errors.Any())
+					{
+						Logger.LogDebug($"Received a response for {action} in {totalTime.TotalSeconds:0.##s} after {attempt} failed attempts: {new AggregateException(errors)}.");
+					}
+					else
+					{
+						Logger.LogDebug($"Received a response for {action} in {totalTime.TotalSeconds:0.##s}.");
+					}
+
+					return response;
+				}
+				catch (Exception e)
```

Comment:
I believe it would be better to catch `HttpRequestException` here as that is the exception that `TorHttpPool` should throw when something is wrong. 

You do not want to catch `OperationCanceledException` for example. Or `InvalidOperationException` for example (that can be found in the source code of `SendAsync` method)

Reply:
I was starting to look into it, and figured I would start with a dumb version and refine it later (see PR description), because I didn't notice tor exceptions were wrapped like this. Looking again, it seems fair to say that this is the only exception type that is reasonable to catch here, would you say that's correct?

Reply:
FWIW if `exceptions.Any()`, but an unrecoverable error occurs, I'm not sure if it's better to just log those like in the success case, and only throw the fatal one, or make an `AggregateException` even though they are qualitatively different. I went with the latter approach for now, but honestly it feels like the less good option the more I think about it... Thoughts?

Reply:
> I was starting to look into it, and figured I would start with a dumb version and refine it later (see PR description), because I didn't notice tor exceptions were wrapped like this. Looking again, it seems fair to say that this is the only exception type that is reasonable to catch here, would you say that's correct?

Well, I designed TorHttpPool to throw only `HttpRequestException`s so that it's similar to `HttpClient`'s behavior (built in .NET). The thing is that starting with dumb version is fine but the problem that arises is that after some time it will be impossible to say why you picked `Exception` and not something else and people will be afraid to modify the code. For this reason, I prefer either a comment or `HttpRequestException`.

Reply:
It's definitely better to be more specific, I was just being lazy.

But my question is, is it really as easy as only `HttpRequestException` or am I missing some other exception class that also might be relevant?

Of course, as you say it's much easier to add specific classes than to remove them, so I already implemented that, and my later comment about the `AggregateException` is already assuming that only `HttpRequestException` is treated as a transient failure

Reply:
> But my question is, is it really as easy as only `HttpRequestException` or am I missing some other exception class that also might be relevant?

Well, the nature of exceptions is that you can never be sure but I'm like 98% confident it's all. If it isn't the fix is easy or we have a bug somewhere, so it makes sense to me to go this way.

Reply:
> FWIW if `exceptions.Any()`, but an unrecoverable error occurs, I'm not sure if it's better to just log those like in the success case, and only throw the fatal one, or make an `AggregateException` even though they are qualitatively different. I went with the latter approach for now, but honestly it feels like the less good option the more I think about it... Thoughts?

TBH, I'm not sure. I would expect the exception to be re-thrown immediately if it is an unrecoverable exception (but how to know which are recoverable and which aren't?).

Reply:
We are defining some exceptions to be recoverable here, by retrying and
hoping to get a better outcome. You've argued correctly and convincingly
that the way i defined it at first was too broad (all expeditions) and that
it's better to start narrowly, and widen the definition as appropriate.

Now that a distinction is being made, the question is if something like
that happens, and then e.g. `InvalidOperationException` is thrown on the
second try, is it better to let that propagate on its own and only log the 1st
exception, or to throw an aggregate exception with both?

Reply:
We are defining some exceptions to be recoverable here, by retrying and
hoping to get a better outcome. You've argued correctly and convincingly
that the way i defined it at first was too broad (all expeditions) and that
it's better to start narrowly, and widen the definition as appropriate.

Now that a distinction is being made, the question is if something like
that happens, and then e.g. `InvalidOperationException` is thrown on the
second try, is it better to let that propagate on its own and only log the 1st
exception, or to throw an aggregate exception with both?

Reply:
> Now that a distinction is being made, the question is if something like
> that happens, and then e.g. `InvalidOperationException` is thrown on the
> second try, is it better to let that propagate on its own and only log the 1st
> exception, or to throw an aggregate exception with both?

I would expect more like to throw a non-aggregate exception. It seems to be easier to handle those exceptions then. Isn't it?

Reply:
I would expect more like to throw a non-aggregate exception. It seems to be easier to handle those exceptions then. Isn't it?

Reply:
yep, that would mean the prior exception would only be logged, but that makes more sense to me because they are not likely to be related in any way.